net: sched: It's ok to preempt coop threads if they're polling

mike-scott opened the pull request:
commit 1856e2206d12 ("kernel/sched: Don't preempt cooperative threads")
changed the behavior of update_cache() in kernel/sched.c so that
high priority threads couldn't be preempted (they are after all
high priority threads).

In kernel/workq.c: work_q_main() the function call to k_queue_get()
can end up using k_poll() if CONFIG_POLL is enabled (such as with
any application with CONFIG_NETWORKING).  Calls to k_poll will
pend threads but for some reason the POLLING state is cleared
right before the pend call.  Then it's cleared again in
signal_poll_event().

Due to the above commit where high priority threads can't be
preempted, we were seeing a random deadlock situation in
work_q_main() where the next high priority thread to be added to
the readyq was a thread that was supposed to be POLLING
(hard to tell when the state was cleared), and when the thread was
unpended, it would return from k_queue_get() with a NULL work.
The work_q_main() function immediate calls k_queue_get() again and
the scheduler immediately unpends the same thread.  Rinse and repeat.

1) Let's keep the POLLING flag in thread_state enabled until it's
cleared in signal_poll_event().
2) In update_cache() we use the POLLING flag to avoid the re-selection
of polling high priority threads.  This should avoid the spin in
work_q_main() since the polling thread will not be unpended until
a signal is hit or it's cancelled.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/8049

Signed-off-by: Michael Scott <michael@opensourcefoundries.com>

mike-scott commented:
Not sure how to update a preempt test for this in tests/kernel/schedule/preempt.

mike-scott commented:
Fixed a minor typo in the commit description.

jukkar commented:
I tried this PR with gptp code. The system works now much better, earlier gPTP was loosing sync after 30min, but now I am not seeing same symptoms even after two hours.

andyross reviewed (changes requested):
OK, so we can't do it with this particular patch as it collides with #8040 which happens to refactor exactly that code.

Let me take a look at k_poll() more carefully.  My understanding was that the POLLING state was used internally to indicate where in the loop the thread was, and that threads sleeping in k_poll() would always be marked PENDING as for any other blocked thread.

If that's not the case, and blocked threads in k_poll() look "runnable", then we almost certainly have more bugs than this!

andyross commented:
(Just to be clear though: this is an amazingly good catch, thanks!)

mike-scott commented:
@andyross I can rebase on top of https://github.com/zephyrproject-rtos/zephyr/pull/8040 if you think that will be merged soon.

andyross commented:
Yeah, unfortunately this isn't a root cause fix.  The POLLING bit is set on a thread only during the critical section at the top of k_poll() where it loops over the handles to see if anything is already flagged.  The only reason it exists is because Ben was concerned that an arbitrary-sized loop exercisable by the app was a latency problem, so he needed to release the irq_lock during execution and then check to see if anything had changed while it was in progress.

The bottom line is that all this patch has done is detect a thread that is stuck spinning in a k_poll() loop and make it magically preemptible.  It's not a root cause fix, the spinning loop is still the real problem.

FWIW: you could probably do this more simply by putting a k_yield() after the irq_unlock() inside the loop at the top of k_poll().  I think that might be more reasonably mergable (though still not a root cause fix!) as it wouldn't change the scheduler semantics and would incur overhead only during polling.  And documenting k_poll() as a yield point would be straightforward and acceptable as it's a blocking call anyway.

mike-scott commented:
@andyross this isn't so much about the POLLING state as it is being able to prempt a high priority thread while it's not ready to run.

We can drop the polling state change in poll.c and change the !_is_thread_polling(_current) addition to _is_thread_ready(_current).   But the essential idea here is the same.  We need to preempt high priority threads when they're in a "not ready" state.   Otherwise the scheduler will spin.

mike-scott committed:
net: sched: It's ok to preempt coop threads if they're not ready

commit 1856e2206d12 ("kernel/sched: Don't preempt cooperative threads")
changed the behavior of update_cache() in kernel/sched.c so that
high priority threads couldn't be preempted (they are after all
high priority threads).

This can cause a random deadlock situation in work_q_main() where the
next high priority thread to be added to the readyq was a thread that
isn't ready to run (i.e. blocked on a signal / timer).  When the thread
was unpended, it would return from k_queue_get() with a NULL work.
The work_q_main() function immediate calls k_queue_get() again and
the scheduler immediately unpends the same thread.  Rinse and repeat.

In update_cache() we should check that a thread is ready before blocking
attempts to preempt it.  This should avoid the spin in work_q_main()
since the thread will not be unpended until it unblocks correctly
or is cancelled.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/8049

Signed-off-by: Michael Scott <michael@opensourcefoundries.com>

andyross commented:
Right, trust me I understand.  But we already have an abstraction for marking a thread when it is "not ready to run", and that's yielding.  The POLLING bit is internal bookeeping for k_poll() to use to finesse synchronization and isn't something we want to be looking at.

Try this change, I'm guessing it's going to be equivalent (if not, try moving the yield call, or even spamming them everywhere -- just don't do it with the lock held, heh):

        --- a/kernel/poll.c
    +++ b/kernel/poll.c
    @@ -222,6 +222,7 @@ int _impl_k_poll(struct k_poll_event *events, int num_events, s32_t timeout)
                            }
                    }
                    irq_unlock(key);
    +               k_yield();
            }
     
            key = irq_lock();

mike-scott commented:
@andyross Explain why we would block a preempt on a high priority thread if the that thread isn't ready to run? (updated the patch with this logic -- removed the k_poll specific changes)

andyross commented:
Becuase that's not how we've documented our scheduling algorithm.  Cooperative threads run until they pend, suspend, yield or abort (except for ISRs obviously, but those are required to return back into the thread before scheduling other user threads).  There's no other "not ready to run" abstraction.

In fact that loop at the top of k_poll)() is not an idle loop!  It's designed to be a fast critical section to catch the race inherent to the API where the events can be signaled before the thread can sleep.

One more time: the root cause here is that the polling thread is not sleeping, it's spinning.  And that's your bug.

mike-scott commented:
The root cause is that we update _kernel.ready_q.cache with the current thread due to new checks in update_cache() and until something else comes along to change that, any call to swap() will re-enable the current thread.  Whether it's ready or not causing the spin.

andyross commented:
Maybe I'm confused then.  How is not rescheduling another thread causing the spin?   What is waking up the work queue thread?  Does the yield patch above not work equivalently?

mike-scott commented:
I think the k_yield() patch is hacky.

At this point in k_poll() we're going to iterate through the events and if there's a valid event which we need to wait for: call register_event().  And then whether or not we've found a valid event we call k_yield outside of the irq_locks so that it can call update_cache in a way that we've decided it's ok to preempt (1).

This seems like a magic fix for this one point in the code.

I still don't understand how we can deny preempt in update_cache() and use the current_thread which could be in a blocked state.

What would happen if we later re-write work_q_main to somehow use a semaphore and the logic in k_sem_take() isn't updated to call k_yield before _pend_current_thread() so that _kernel.ready_q.cache is updated?

mike-scott commented:
Just because a thread is PENDING doesn't stop all of the arch swap() functions from using it if ready_q.cache points to it:
ARM: https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/arm/core/swap_helper.S#L111
ARC: https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/arc/core/swap.S#L102
NIOS2: https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/nios2/core/swap.S#L67
POSIX: https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/posix/core/swap.c#L65
X86: https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/x86/core/swap.S#L150

mike-scott commented:
IMHO: we cannot/will not point ready_q.cache at any thread which is "not ready to run".

mike-scott commented:
Btw, I understand how _THREAD_POLLING is used now.  I dropped it a while back.

andyross commented:
It does, though, because a pended thread isn't in the ready queue to begin with and so can't possibly be tested for placement in the cache pointer.

I still think you're misunderstanding this: a thread in that loop at the top of k_poll() **is absolutely "ready to run"**.  It's in the process of deciding whether or not any of the poll events it's been given have already been signaled, and if they aren't, it will then go to sleep.  All your patch does is make threads in this relatively tight loop look "preemptible" even if they are at a cooperative priority -- it's semantically equivalent to just putting a yield between every (unlocked) statement.

(FWIW: that whole "cache" thing is an optimization to allow for simpler per-arch assembly: all is does is store the pre-selected "best next thread to run" based on what is happening in the queue state, which is why the preemption logic lives there and not at the actual point of preemption in _Swap() and/or ISR return).

mike-scott commented:
There's no logic in update_cache() to stop from assigning the _current thread in whatever state its in to ready_q.cache:
https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/sched.c#L175

andyross commented:
Oh... heh, now I'm looking at your patch again.  Sorry, I was still picturing the polling change.  That... looks really wrong if this is fixing your problem and implies a deeper scheduler bug.  Unready threads should never have gotten to that level of logic.

Can you please apply the change in #8040 (which refactors this code into a should_preempt() predicate) and see if the symptom still occurs, and if it is slot your ready() test in there?

mike-scott commented:
No logic to check if the current thread is in the ready queue that is.

mike-scott commented:
With the latest scheduler changes merged, this patch has been integrated into:
https://github.com/zephyrproject-rtos/zephyr/pull/8077
